host/logmux/sink.go

```diff
@@ -0,0 +1,604 @@
+	s.shutdownOnce.Do(func() { close(s.shutdownCh) })
+}
+
+func (s *LogAggregatorSink) ShutdownCh() chan struct{} {
+	return s.shutdownCh
+}
+
+// SyslogSink is a flexible sink that can connect to TCP/TLS endpoints that use syslog framing.
+// The body of the message can be customised using a template.
+type SyslogSink struct {
+	sm *SinkManager
+
+	id     string
+	url    string
+	prefix string
+
+	mtx          sync.RWMutex
+	cache        *lru.Cache
+	cursor       *utils.HostCursor
+	template     *template.Template
+	conn         net.Conn
+	shutdownOnce sync.Once
+	shutdownCh   chan struct{}
+}
+
+func NewSyslogSink(sm *SinkManager, info *SinkInfo) (sink *SyslogSink, err error) {
+	cfg := &ct.SyslogSinkConfig{}
+	if err := json.Unmarshal(info.Config, cfg); err != nil {
+		return nil, err
+	}
+	var t *template.Template
+	var cache *lru.Cache
+	if cfg.Prefix != "" {
+		// Initialse the template cache
+		cache = lru.New(1000) // TODO(jpg): Consider configurable?
+		t, err = template.New("").Parse(cfg.Prefix)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &SyslogSink{
+		sm:         sm,
+		id:         info.ID,
+		url:        cfg.URL,
+		prefix:     cfg.Prefix,
+		cache:      cache,
+		template:   t,
+		cursor:     info.Cursor,
+		shutdownCh: make(chan struct{}),
+	}, nil
+}
+
+func (s *SyslogSink) Info() *SinkInfo {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	config, _ := json.Marshal(ct.SyslogSinkConfig{URL: s.url, Prefix: s.prefix})
+	return &SinkInfo{
+		ID:     s.id,
+		Kind:   ct.SinkKindSyslog,
+		Config: config,
+		Cursor: s.cursor,
+	}
+}
+
+func (s *SyslogSink) Connect() error {
+	u, err := url.Parse(s.url)
+	if err != nil {
+		return err
+	}
+	host, port, _ := net.SplitHostPort(u.Host)
+	if port == "" {
+		port = "514"
+	}
+	addr := net.JoinHostPort(host, port)
+	var conn net.Conn
+	switch u.Scheme {
+	case "syslog":
+		conn, err = net.Dial("tcp", addr)
+	case "syslog+tls":
+		tlsConfig := tlsconfig.SecureCiphers(&tls.Config{})
+		conn, err = tls.Dial("tcp", addr, tlsConfig)
+	default:
+		return fmt.Errorf("unknown protocol %s", u.Scheme)
+	}
+	if err != nil {
+		return err
+	}
+	s.conn = conn
+	return nil
+}
+
+func (s *SyslogSink) GetCursor(_ string) (*utils.HostCursor, error) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	return s.cursor, nil
+}
+
+// Parses JobID and ProcType from header ProcID field
+// Always returns JobID, returns ProcType if available
+func parseProcID(procID []byte) (string, string) {
+	procTypeID := strings.SplitN(string(procID), ".", 2)
+	if len(procTypeID) == 2 {
+		return procTypeID[1], procTypeID[0]
+	}
+	return procTypeID[0], ""
+}
+
+var msgSep = []byte{' '}
+
+func (s *SyslogSink) Write(m message) error {
+	if s.template != nil {
+		// Lookup rendered template from cache
+		var prefix []byte
+		if cached, ok := s.cache.Get(m.Message.ProcID); ok {
+			if p, ok := cached.([]byte); ok {
+				prefix = p
+			}
+		}
+		// If not in the cache execute the template and cache the result
+		if len(prefix) == 0 {
```

Comment:
This should probably check if the prefix is nil, as a template could return a zero-length result.